lib/locations/locations.go

```diff
@@ -123,23 +131,42 @@ func defaultConfigDir() string {
 		return filepath.Join(os.Getenv("AppData"), "Syncthing")
 
 	case "darwin":
-		dir, err := fs.ExpandTilde("~/Library/Application Support/Syncthing")
-		if err != nil {
-			fmt.Println(err)
-			panic("Failed to get default config dir")
-		}
-		return dir
+		return filepath.Join(home, "Library/Application Support/Syncthing")
 
 	default:
 		if xdgCfg := os.Getenv("XDG_CONFIG_HOME"); xdgCfg != "" {
 			return filepath.Join(xdgCfg, "syncthing")
 		}
-		dir, err := fs.ExpandTilde("~/.config/syncthing")
-		if err != nil {
-			fmt.Println(err)
-			panic("Failed to get default config dir")
+		return filepath.Join(home, ".config/syncthing")
+	}
+}
+
+// defaultDataDir returns the default data directory, which usually is the
+// config directory but might be something else.
+func defaultDataDir(home, config string) string {
+	switch runtime.GOOS {
+	case "windows", "darwin":
+		return config
+
+	default:
+		// If a database exists at the "normal" location, use that anyway.
+		if _, err := os.Lstat(filepath.Join(home, "syncthing", databaseDirname)); err == nil {
+			return config
+		}
+		// Always use this env var, as it's explicitly set by the user
+		if xdgHome := os.Getenv("XDG_DATA_HOME"); xdgHome != "" {
+			return xdgHome
```

Comment:
So `XDG_DATA_HOME` here will be something like `~/.local/share` which is not where we should store the database, we should use a `syncthing` subdir? (Same for xdgDefault stuff below I think)